pkg/cli/cli.go

```diff
@@ -983,23 +1000,166 @@ func printConnection(c *command, cvd RemoteCVDLocator, status ConnStatus) {
-	}
+func buildAgentCmdArgs(flags *ConnectFlags, device string, connectAgent string) []string {
+	args := []string{connectAgent, device}
 	return append(args, flags.AsArgs()...)
 }
 
-// Handler for the agent command. This is not meant to be called by the user
-// directly, but instead is started by the open command.
+// Return the ADB address of remote Cuttlefish instance as [host]:port.
+func findDeviceAdbAddress(service client.Service, controlDir, hostname, device string) (string, error) {
+	// Retrieving host to get IP address of host
+	host, err := findHost(service, hostname)
+	if err != nil {
+		return "", fmt.Errorf("failed to find host")
+	}
+	if host.Docker == nil {
+		return "", errors.New("instance type should be Docker")
+	}
+
+	// Retriving ADB port of Cuttlefish device
+	cvd, err := findCVD(service, controlDir, hostname, device)
+	if err != nil {
+		return "", fmt.Errorf("failed to find cvd: %w", err)
+	}
+	_, port, err := net.SplitHostPort(cvd.ADBSerial)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse port from ADB serial: %w", err)
+	}
+	if port == "" {
+		return "", errors.New("failed to find port")
+	}
+
+	return net.JoinHostPort(host.Docker.IPAddress, port), nil
+}
+
+// Letting the process be a proxy server for establishing the connection.
+func forwardProxy(controlDir, proxyAddr, adbAddr string, adbServerProxy ADBServerProxy, host, device string) error {
+	// Make connection towards [host_ip_address]:[cuttlefish_instance_adb_port]
+	var dialer proxy.Dialer
+	if proxyAddr != "" {
+		proxyUrl, err := url.Parse(proxyAddr)
+		if err != nil {
+			return fmt.Errorf("failed to parse proxy URL: %w", err)
+		}
+		if proxyUrl.Scheme != "socks5" {
+			return fmt.Errorf("scheme of proxy URL is not socks5. actual: %s", proxyUrl.Scheme)
+		}
+		dialer, err = proxy.SOCKS5("tcp", proxyUrl.Host, nil, nil)
+		if err != nil {
+			return fmt.Errorf("failed to create proxy dialer: %w", err)
+		}
+	} else {
+		dialer = proxy.Direct
+	}
+	remoteConn, err := dialer.Dial("tcp", adbAddr)
+	if err != nil {
+		return fmt.Errorf("failed to dial remote port: %w", err)
+	}
+	defer remoteConn.Close()
+
+	// Create a file socket to establish ADB connection
+	socketPath := GetProxySocketPath(controlDir, host, device)
+	localListener, err := net.Listen("unix", socketPath)
+	if err != nil {
+		return fmt.Errorf("failed to create unix socket: %w", err)
+	}
+	defer localListener.Close()
+
+	// Enroll signal channel to handle SIGINT and SIGTERM.
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
+
+	// Initialize ADB connection
+	errorCh := make(chan error)
+	go func() {
+		if err := adbServerProxy.ConnectWithLocalFileSystem(socketPath); err != nil {
+			errorCh <- fmt.Errorf("failed to connect adb: %w", err)
+		}
+	}()
+
+	// Listener is accepted by ADB connection.
+	startLocalToRemoteCh := make(chan net.Conn)
+	startRemoteToLocalCh := make(chan net.Conn)
+	go func() {
+		localConn, err := localListener.Accept()
```

Comment:
wondering whether the `Accept` call can be done in the same `goroutine` to avoid the `startLocalToRemoteCh` and `startRemoteToLocalCh` channels.

Reply:
Current execution flow is,
1. Start running `adbServerProxy.ConnectWithLocalFileSystem(socketPath)`.
2. Blocked by `localListener.Accept()` is disappeared.
3. Start running `io.copy` for both direction.
4. Finish running `adbServerProxy.ConnectWithLocalFileSystem(socketPath)`.

The problem is, both `io.copy` are intended to run forever while keeping the ADB connection. So each `io.copy` needs its own goroutine. We can reduce one channel, but we cannot remove both connection. But in the perspective of readability, I think reducing only one channel wouldn't be the better one. Please reopen this if you have another idea.

Reply:
Sorry, I miscommunicated. 

I wanted to know whether we could get rid of the goroutine for `localConn, err := localListener.Accept()`, this way the `localConn` could be used in the following `io.Copy` goroutines removing the need of having two different channels to send the `localConn` instance over.

If we do need the `goroutine`, can we have one channel (`localConnCh`)  and send the `localConn` instance twice?

Reply:
IMO, we cannot remove the goroutine for `localListener.Accept()`. `. While waiting for `localListener.Accept()` w/o goroutine, we cannot wait `errCh` sent from `adbServerProxy.ConnectWithLocalFileSystem(socketPath)`.

And about having one channel(`localConnCh`), I already tried it before. However in that trial, one of 2 goroutines couldn't get `localConn` properly. That's why I defined 2 channels for transmitting `localConn`.